src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs

```diff
@@ -29,7 +29,9 @@ internal abstract class DictionaryDefaultConverter<TCollection, TKey, TValue>
         /// </summary>
         protected virtual void CreateCollection(ref Utf8JsonReader reader, ref ReadStack state) { }
 
-        internal override Type ElementType => typeof(TValue);
+        private static Type s_valueType = typeof(TValue);
+
+        internal override Type ElementType => s_valueType;
```

Comment:
I thought this was called during throughput time: a repeated call to `typeof()` at (de)serialization time is more costly than caching the type. https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMIEDeOB1BADgE5gBuAhsAKZHoBsBAKgE9anAM4B9MADtgg4QQC8BYEPYB7AGYAKKcACUAbio0j1fPxUEA4u2ABJabPabdCgHxKVG7dIMmCfs0crG3sZFRIWAGMAC3YYZzcCcR1HQ2wAXyA===

Since `ElementType` is only used during warm-up, I'll revert to make this cleaner(?).

Reply:
I thought code paths that use `typeof(....)` get optimized. What is the generated disassembly in your examples (sharplab can't show it)? Also, use release.

Reply:
Perhaps. You should be able to see it now: https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJmADgE5gBuAhsAKaYAqAnmUwM4D6YAdsDYdMAXkzB2DAPYAzABT9gASgDcxUupIYWkzAHEGwAJIChDOUtEA+cZNkKBqzZmfaz+wycGTctAMYAFgwAJhbWmDyKZmooAL5AA

Reply:
I doubt that extra mov you see in the simplified example is meaningful.

An actual usage is more important to look at:
[sharplab.io](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4CS51ATgPYAOAyswG4UDG1CAG50WXARLkqtAMIsobCgBtmXJrwHDRODgAswTNgBkwAIyJlKNEWlEBmXAmwzsAb3TZP2NkwrcwwNTYACoAnmxBEAD6FORhEdgAvNjA4dQsAGYAFKahgQCUNl7YHl5wDvFBAOLUwAzAlVn5SQB8KWmZOXnUhaWefbgVadg1dXFpMmB8utQAJk2t2NGxDWlFXgMA2gBSFMA1pMz8WakRnbkF+QC6A+XYK9hcAI7E1KQCcgoG1MEsmAA8wRaWQ4bDApEBbQgYNIABpsIRqGBZgB5UhKUKg8GQ7AsYAzJj5AbFYoAdwJQWC2BA2DonzBTDMKkhA3caBJXgoGWwJw62WCzUSyVO6WyFx6RPZHJJcAA7I9qC83h95AyfiwssTpcUmNRuQBVUgQMAZWgAQQggPh4uBuu5AFlqFAWExQvaDBB9Ep8KNEabde9qFloeD8vlYVrtZ4Q6R8EY3gBzfERqVRkl27CG42m/AWq3YG1ZDOO52u91MT1gb2+vXMZVBvEEsMptMcxvMOOJ/G9WypklcnkizoCpLCvlZaYGMORmXy56vQP076/TV91vYDNZk3my3BeGTpi2vXYEsut0er0+2p+uuB4Mw5sz7UxzukJO6Fvrrybo3b3O7/d9EPItj1PMsLyrK9gBvAMBCydtCXDJ9pQQ193x7Gc5QVJVF1VZcNWLJ0z3LStq2vWtYKDGNwyWGE0OTDdQKI8CK0vGt/XreD8WYGjUPjN9u3WTwAF8Bi2XZ9jeI4+F5M4xW6MMbj7O4Hnnesl11X4kEhEEYRxGN4URZE0QxLEISBXFuMJZDPHJOsQmpWkNKZagWT7NltQHWTRSyEchRGWp6kaac12lLC1Nwr5NI1Gz02PLcczzPcC26I8HWY89WMg9jbzg6ikNCtMX3499Py/RiDV/RKAJSwI0pPDKSLY8iOLvBDH0KqM+K7XQMM6ryh35QUxzkicgJC9dwsVBcVSi9VV3Kzwf2zHd8wPeqwMy0ioJgzj8rKr9ip6g712Wv8ksAgwNsaiCyOgijOPagrFssgl6N6mxMLnab1Lw6KQPS0stua+7Wryh94SOgSPwqhqgaa7KWtyhsrN4qz3p7YpRL7cS9gOaTvPOBTrluBxVJ+yK1V+BAdLM/SYUMpFUXRTE9Is9rYrs3UHJpOk/pctzig86UBvHPzkmWcYIklSbvpw2aqZizrtTO6r80LQj4dunaHrvfbYuKKHSoN794qq1bko1pitayu7dra1HnsW7rob6zzuUJobR3aUaDwm1spvl6gNPmk2lrNlb/zWoDrpt7acso+9Qyd8qjeTMPYYSi3LuAzXiO1hPHsdk7Wxd9DPs6wOZuDv75rzlj46RxPqMhuiSoY+vgcR0Hka4pt4TLwSBmx4pcckw5fBkwaukuJTihU8hsOrkPfjsWm2ahBmESZkzWexdnUc5ikeac/nTGZIFWRnUXRvFgKxlWCJJmmOYmhlgO5eX2uVwz1Xs9q6gsd862x1mDKiD4U6HTbsdX+Edzo1StoDYBjce6JyeiXNMg8ProGvh7aed9p5+3fmmKuv05o/2VtKP+UdkrrQBnDZBIN7bg2ThgqMacPywMqpHC62A6GdwRnbXWcF0EZywW7MKn8yGK3oZtQRoDe4t1ouCd68IBEFybkXfur0Ozt2wX2EeGw+ybEdPiFgsw6AKCUFkUxuhzGWLYEoFEbBKAsCNPgAAciwBgShYixATCTPsPg/ABCCHgAAbPcReEUFb4XoeKAsSB4QPCUEk2GCTTAIGSYvJQCBiFuC+tgTAQlsCGP6MY2x9irE2NqHYixVjnGuPcV4nxfi3yBOKME/wgRxCRPJkHFeBFjwHkSdk4A2BUlqOGUBAsWSonjNyfk4WHIsLFOHmJCptSqmOJqWY+pjjGkUDcRATx3j0RtICXPLwXTQm9PmUvaR+EdIZipKYNJKS0kvNmWMiZeSr6V3lGsgx6BhJAA)

The assembly of implementation 0 and 1 are identical, which use `typeof` directly and `GetIntType()`.

Implementation 2 and 3 have more code when accessing the field and `GetIntTypeCached`. Try it out, locally, with `static` (and readonly) and see if that makes a difference. Similarly, investigate if used as `virtual` makes caching more useful.